### PR TITLE
HTTP connection pooling

### PR DESCRIPTION
**Pull request by andreachild**
Changed connection pooling to reflect having switched from long lived web socket connections to short lived http connections:
- sped up obtaining a new connection by removing method-level synchronization of getAvailableConnection and introducing a separate set for available connections
- introduced cleanup of idle connections to release resources
- removed minSize since idle connections are now cleaned up and the pool can become empty if there is no traffic
- increased maxSize default to 128
- increased core pool size of ScheduledThreadPoolExecutor that is used for creating new connections to number of available processors
- added logging to help troubleshoot connection related integration tests failures

ProfilingApplication throughput results of master-http prior to these changes (with small min pool size of 50):
```
[ec2-user@ip-xxxxx bin]$ ./profile-driver.sh parallelism 16 executions 5 requests 1000000 minConnectionPoolSize 50 maxConnectionPoolSize 5000 host xxxxx
---------------------THROUGHPUT TEST SELECTED---------------------
---------------------------WARMUP CYCLE---------------------------
[warmup-1] requests: 1000 | time(s): 0.874164166    | req/sec: 1144    | too slow: 0
[warmup-2] requests: 1000 | time(s): 0.80377202     | req/sec: 1244    | too slow: 0
[warmup-3] requests: 1000 | time(s): 0.797980361    | req/sec: 1253    | too slow: 0
[warmup-4] requests: 1000 | time(s): 0.798658453    | req/sec: 1252    | too slow: 0
[warmup-5] requests: 1000 | time(s): 0.798385122    | req/sec: 1253    | too slow: 0
----------------------------TEST CYCLE----------------------------
[test-1]   requests: 1000000 | time(s): 48.756697463   | req/sec: 20510   | too slow: 1
[test-2]   requests: 1000000 | time(s): 48.614801918   | req/sec: 20570   | too slow: 0
[test-3]   requests: 1000000 | time(s): 48.647116064   | req/sec: 20556   | too slow: 0
[test-4]   requests: 1000000 | time(s): 48.577865007   | req/sec: 20586   | too slow: 0
[test-5]   requests: 1000000 | time(s): 48.646597705   | req/sec: 20556   | too slow: 0
avg req/sec: 20555
```

ProfilingApplication throughput results after these changes are applied (no more min pool size but connection pool is created initially with 1 connection at startup):
```
[ec2-user@ip-xxxxx bin]$ ./profile-driver.sh parallelism 16 executions 5 requests 1000000 maxConnectionPoolSize 5000 host xxxxx
---------------------THROUGHPUT TEST SELECTED---------------------
---------------------------WARMUP CYCLE---------------------------
[warmup-1] requests: 1000 | time(s): 0.829135665    | req/sec: 1206    | too slow: 0
[warmup-2] requests: 1000 | time(s): 0.75848205     | req/sec: 1318    | too slow: 0
[warmup-3] requests: 1000 | time(s): 0.853421271    | req/sec: 1172    | too slow: 0
[warmup-4] requests: 1000 | time(s): 0.755844005    | req/sec: 1323    | too slow: 0
[warmup-5] requests: 1000 | time(s): 0.800583459    | req/sec: 1249    | too slow: 0
----------------------------TEST CYCLE----------------------------
[test-1]   requests: 1000000 | time(s): 31.881409634   | req/sec: 31366   | too slow: 0
[test-2]   requests: 1000000 | time(s): 31.6027552     | req/sec: 31643   | too slow: 0
[test-3]   requests: 1000000 | time(s): 32.396470504   | req/sec: 30868   | too slow: 0
[test-4]   requests: 1000000 | time(s): 32.165162331   | req/sec: 31090   | too slow: 0
[test-5]   requests: 1000000 | time(s): 31.820634099   | req/sec: 31426   | too slow: 0
avg req/sec: 31278
```

JProfiler screenshot of hotspots prior to these changes:
Screenshot 2024-10-22 at 2 08 39 PM

JProfiler screenshot of hotspots after these changes:
Screenshot 2024-10-30 at 10 20 12 AM

### CODE REVIEW

**Comment by kenhuuu** [comment]
What happens in the case where a query is sent to the server and the server needs more processing time than the idle connection timeout before sending back results? I think a test case might need to be added for this scenario.

File: gremlin-server/src/test/java/org/apache/tinkerpop/gremlin/driver/ClientConnectionIntegrateTest.java [hunk 1]
```diff
@@ -292,6 +296,55 @@ public void shouldSucceedWithJitteryConnection() throws Exception {
...
+
+        } finally {
+            executorServiceForTesting.shutdown();
+            client.close();
+            cluster.close();
+        }
+    }
+
+    private static void chooseConnections(int operations, Client.ClusteredClient client, ExecutorService executorServiceForTesting) throws InterruptedException {
+        final RequestMessage.Builder request = client.buildMessage(RequestMessage.build("1+1"));
```
[the remark this answers is not in the document]
Comment (andreachild) [reply gremlin-server/src/test/java/org/apache/tinkerpop/gremlin/driver/ClientConnectionIntegrateTest.java]: Changed to `g.inject(1)`

**Comment by andreachild** [comment]
> What happens in the case where a query is sent to the server and the server needs more processing time than the idle connection timeout before sending back results? I think a test case might need to be added for this scenario.

@kenhuuu I added a change to address this - I changed SslCheckHandler to a generic 'channel closed' handler which detects if it closed due to idle timeout and then throws an appropriate error message. Added integration for this scenario as well.

File: gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/ConnectionPool.java [hunk 2]
```diff
@@ -59,8 +58,8 @@ final class ConnectionPool {
     private final Client client;
     private final List<Connection> connections;
     private final AtomicInteger open;
+    private final Set<Connection> availableConnections = new CopyOnWriteArraySet<>();
```
Comment (Cole-Greer) [inline gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/ConnectionPool.java:61]: CopyOnWriteArraySet does have some performance implications for applications which have a lot of insertions and removals although many alternatives carry their own implications. Have you done any profiling which may show whether this is a real concern?

File: gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/Connection.java [hunk 3]
```diff
@@ -111,15 +112,15 @@ public Connection(final URI uri, final ConnectionPool pool) throws ConnectionExc
                 // close did not come from the client side. it means the server closed the channel for some reason.
                 // it's important to distinguish that difference in debugging
                 if (thisConnection.closeFuture.get() == null) {
-                    logger.error(String.format(
-                            "Server closed the Connection on channel %s - scheduling removal from %s",
+                    logger.info(String.format(
```
Comment (kenhuuu) [inline gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/Connection.java:115]: Why is this changed from error to info? Because this is still used to capture cases where the server shut the connection, it should probably still be a higher value like "error"?

File: gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/ConnectionPool.java [hunk 4]
```diff
@@ -369,15 +345,6 @@ private boolean addConnectionIfUnderMaximum() {
     }
 
     private boolean destroyConnection(final Connection connection) {
```
Comment (kenhuuu) [inline gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/ConnectionPool.java:347]: Nit: this and definitelyDestroyConnection() could get merged into single method

File: gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/handler/IdleConnectionHandler.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    public static final AttributeKey<IdleStateEvent> IDLE_STATE_EVENT = AttributeKey.valueOf("idleStateEvent");
+
+    /**
+     * Detects if an {@link IdleStateEvent} has occurred and closes the channel.
+     */
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
+        if (evt instanceof IdleStateEvent) {
+            final IdleStateEvent e = (IdleStateEvent) evt;
+            logger.info("Detected IdleStateEvent {} - closing channel", e);
```
Comment (kenhuuu) [inline gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/handler/IdleConnectionHandler.java:44]: Nit: log the channel ID here.

**Comment by kenhuuu** [comment]
VOTE +1, pending small nits.

File: gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/ConnectionPool.java (hunk 2 above)
Comment (andreachild) [reply gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/ConnectionPool.java]: I've changed it to ConcurrentLinkedQueue which will be faster for insert and removals and also keep older connections at the head so that they are used next, which is consistent with previous behaviour.

File: gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/Connection.java (hunk 3 above)
Comment (andreachild) [reply gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/Connection.java]: Added check that if the connection was closed and it wasn't due to idle timeout, then to log this at error level.

File: gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/ConnectionPool.java (hunk 4 above)
Comment (andreachild) [reply gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/ConnectionPool.java]: Merged

File: gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/handler/IdleConnectionHandler.java (hunk 5 above)
Comment (andreachild) [reply gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/handler/IdleConnectionHandler.java]: Logged channel id

Comment (kenhuuu) [reply gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/handler/IdleConnectionHandler.java]: I should have clarified, in `ConnectionPool` we normally log the short id so that is what should probably be used here too  (`id().asShortText()`).

Comment (andreachild) [reply gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/handler/IdleConnectionHandler.java]: Changed to `asShortText()`

**Comment by Cole-Greer** [comment]
Thanks @andreachild, looks great.

VOTE +1

**Comment by xiazcy** [comment]
VOTE +1. Thanks for the change! If you could post the performance results/comparisons it would be great as well.

**Comment by andreachild** [comment]
> VOTE +1. Thanks for the change! If you could post the performance results/comparisons it would be great as well.

Posted throughput results and profiler screenshots
